### ISSUE TITLE
Simplify and Improve Host Services Developer Experience

### ISSUE THREAD

**Issue by autodidaddict** [synadia-io/nex#332]
### Proposed change

* No longer use a separate connection for triggers, e.g. there's one host services connection to rule them all
* Allow the host services client API to select a different resource, e.g. use a different KV or Object Store than the one that host services auto-provisions
* related to the above: don't auto-provision a host services resource unless it's needed. This means that if the only resource a function or service accesses is a pre-existing one, host services will have not provisioned anything new

This new API might look like this:
```javascript
// Use the default auto-provisioned host services bucket
 this.hostServices.kv.set('hello', payload);
// Use a bucket not managed by host services
this.hostServices.kv.set('hello', payload, "MyBucket");
```

### Use case

After working with host services in a number of real scenarios, we realized that the developer experience was too high friction and could be simplified greatly and make it easier to create new functions that utilize existing app resources.

**Comment by bruth** [comment]
Some thoughts:

> don't auto-provision a host services resource unless it's needed
Two concerns for this happening at runtime are:

- High latency for the first message due to taking some time to provision the resource (not a huge deal, but not a great initial experience)
- Inability to provision the resource due to limits or some unanticipated error

> After working with host services in a number of real scenarios, we realized that the developer experience was too high friction and could be simplified greatly and make it easier to create new functions that utilize existing app resources

Given this statement, what would be the subset of use cases that _prefer_ an auto-provisioned set of resources? If it is a significant minority, is it worth the effort to support this? Or should the expectation be that resources need to be pre-provisioned and the `hostServices` interface is a simplified client for access them?

Another thought, reducing `hostServices` to just `services` or removing that layer altogether, e.g. `this.kv.set(...)` could be viable if it wouldn't pollute other properties on the injected context.

For the API itself, having the bucket as the third parameter is a bit awkward IMO since, for example the NATS CLI takes bucket -> key -> payload. Given the new scope of the design could accessors be used?

```
// Access a named bucket.
this.kv['MyBucket'].set(...)

// or set to a variable for multiple uses..
const kv = this.kv['MyBucket']
kv.get(...)
kv.set(...)
```

If the auto-provisioned option is still available, given it's javascript, methods could still exist on `this.kv` to delegate to the internal one.

**Comment by autodidaddict** [comment]
Agreed. We could make auto provision something that is used for dev and testing and assume pre existing for "real" ops

**Comment by MikeHawkesCapventis** [comment]
That's nice - it also gives the ability to have scripts utilising things like config stored in sharable KV or Object buckets. Obviously, it'd be up to the script to manage any potential concurrency issues, if used in that way.

**Comment by kthomas** [comment]
FWIW I think the best DX would be to support `jit_provisioning` of resources as a flag in the host services configuration, default `true`.

**Comment by autodidaddict** [comment]
closed by #334
